bundle/statemgmt/state_pull.go

```diff
@@ -4,129 +4,262 @@ import (
+func filerRead(ctx context.Context, f filer.Filer, path string, isDirect bool) *state {
+	state, err := _filerRead(ctx, f, path)
 	if err != nil {
-		return diag.FromErr(err)
+		logdiag.LogError(ctx, fmt.Errorf("reading %s: %w", path, err))
+	} else if state != nil {
+		log.Debugf(ctx, "read %s: %s", path, state.String())
+		state.isDirect = isDirect
 	}
+	return state
+}
 
-	// Case: Remote state file does not exist. In this case we fallback to using the
-	// local resources state. This allows users to change the "root_path" their bundle is
-	// configured with.
-	remoteState, remoteContent, err := l.remoteState(ctx, b)
-	if errors.Is(err, fs.ErrNotExist) {
-		log.Infof(ctx, "Remote state file does not exist. Using local resources state.")
-		return nil
+func PullResourcesState(ctx context.Context, b *bundle.Bundle, alwaysPull AlwaysPull) context.Context {
+	_, localPathDirect := b.StateFilenameDirect(ctx)
+	_, localPathTerraform := b.StateFilenameTerraform(ctx)
+
+	states := readStates(ctx, b, alwaysPull)
+
+	if logdiag.HasError(ctx) {
+		return ctx
 	}
-	if err != nil {
-		return diag.Errorf("failed to read remote state file: %v", err)
+
+	var winner *state
+
+	if len(states) == 0 {
+		// no local or remote state; set b.DirectDeployment based on env vars
+		isDirect, err := getDirectDeploymentEnv(ctx)
+		if err != nil {
+			logdiag.LogError(ctx, err)
+			return nil
+		}
+		b.DirectDeployment = &isDirect
+	} else {
+		winner = states[len(states)-1]
+		b.DirectDeployment = ptrBool(winner.isDirect)
+	}
+
+	engine := "direct"
+
+	if !*b.DirectDeployment {
+		engine = "terraform"
 	}
 
-	// Expected invariant: remote state file should have a lineage UUID. Error
-	// if that's not the case.
-	if remoteState.Lineage == "" {
-		return diag.Errorf("remote state file does not have a lineage")
+	// Set the engine in the user agent
+	ctx = useragent.InContext(ctx, "engine", engine)
+
+	if winner == nil {
+		log.Infof(ctx, "No existing resource state found")
+		return ctx
 	}
 
-	// Case: Local state file does not exist. In this case we should rely on the remote state file.
-	localState, err := l.localState(ctx, b)
-	if errors.Is(err, fs.ErrNotExist) {
-		log.Infof(ctx, "Local state file does not exist. Using remote resources state.")
-		err := os.WriteFile(localStatePath, remoteContent, 0o600)
-		return diag.FromErr(err)
+	var stateStrs []string
+	for _, state := range states {
+		stateStrs = append(stateStrs, state.String())
 	}
-	if err != nil {
-		return diag.Errorf("failed to read local state file: %v", err)
+
+	log.Infof(ctx, "Available resource state files (from least to most preferred): %s", strings.Join(stateStrs, ", "))
+
+	var lastLineage *state
+
+	for _, state := range states {
+		if lastLineage == nil {
+			lastLineage = state
+		} else if lastLineage.Lineage != state.Lineage {
+			logdiag.LogError(ctx, fmt.Errorf("lineage mismatch in state files: %s vs %s", lastLineage.String(), state.String()))
```

Comment:
I assume this means during migration the lineage from TF will be carried over to direct state?

Reply:
yes